packages/react/src/popover/store.ts

```diff
@@ -0,0 +1,111 @@
+import { Store, createSelector } from '@base-ui-components/utils/store';
+import { type InteractionType } from '@base-ui-components/utils/useEnhancedClickHandler';
+import { type FloatingRootContext } from '../floating-ui-react';
+import { type TransitionStatus } from '../utils/useTransitionStatus';
+import { type PopoverOpenChangeReason } from './root/PopoverRootContext';
+import { type HTMLProps } from '../utils/types';
+import { getEmptyContext } from '../floating-ui-react/hooks/useFloatingRootContext';
+
+const EMPTY_OBJ = {};
+
+export type State = {
+  modal: boolean | 'trap-focus';
+
+  open: boolean;
+  activeTriggerElement: HTMLElement | null;
+  positionerElement: HTMLElement | null;
+  popupElement: HTMLElement | null;
+  triggers: Map<HTMLElement, (() => unknown) | undefined>;
+
+  instantType: 'dismiss' | 'click' | undefined;
+  transitionStatus: TransitionStatus;
+  openMethod: InteractionType | null;
+  openReason: PopoverOpenChangeReason | null;
+
+  titleId: string | undefined;
+  descriptionId: string | undefined;
+
+  floatingRootContext: FloatingRootContext;
+
+  payload: unknown | undefined;
+
+  activeTriggerProps: HTMLProps;
+  inactiveTriggerProps: HTMLProps;
+  popupProps: HTMLProps;
+  stickIfOpen: boolean;
+};
+
+function createInitialState<Payload>(): State {
+  return {
+    open: false,
+    modal: false,
+    activeTriggerElement: null,
+    positionerElement: null,
+    popupElement: null,
+    triggers: new Map<HTMLElement, (() => Payload) | undefined>(),
+    instantType: undefined,
+    transitionStatus: 'idle',
+    openMethod: null,
+    openReason: null,
+    titleId: undefined,
+    descriptionId: undefined,
+    floatingRootContext: getEmptyContext(),
+    payload: undefined,
+    activeTriggerProps: EMPTY_OBJ as HTMLProps,
+    inactiveTriggerProps: EMPTY_OBJ as HTMLProps,
+    popupProps: EMPTY_OBJ as HTMLProps,
+    stickIfOpen: true,
+  };
+}
+
+export class PopoverStore<Payload = undefined> extends Store<State> {
+  constructor(initialState?: Partial<State>) {
+    super({ ...createInitialState<Payload>(), ...initialState });
+  }
+
+  registerTrigger(triggerElement: HTMLElement, getPayload?: () => Payload) {
+    const triggers = this.state.triggers;
+    triggers.set(triggerElement, getPayload);
+
+    this.set('triggers', new Map(triggers));
+  }
+
+  unregisterTrigger(triggerElement: HTMLElement) {
+    const triggers = this.state.triggers;
+    triggers.delete(triggerElement);
+
+    this.set('triggers', new Map(triggers));
+  }
```

Comment:
To avoid weird bugs, it would be preferable to create the new map before modifying it, rather than modifying the old one.
```suggestion
  registerTrigger(triggerElement: HTMLElement, getPayload?: () => Payload) {
    const triggers = new Map(this.state.triggers);
    triggers.set(triggerElement, getPayload);
    this.set('triggers', triggers);
  }

  unregisterTrigger(triggerElement: HTMLElement) {
    const triggers = new Map(this.state.triggers);
    triggers.delete(triggerElement);
    this.set('triggers', triggers);
  }
```

Reply:
Good point, thanks. Updated.